Change the icloud device_tracker platform into a new component

Bart274 opened the pull request:
**Description:**
This PR changes the old icloud device_tracker into a new platform.
This adds the following features to the new platform:
- Each device has it's own dynamic interval instead of one fixed interval for all devices linked to one account
- This adds the ability to ignore specific devices that you don't want to track, so their battery isn't drained
- This adds the ability to update the devices on a fixed time no matter what the interval is at that time
- This adds 3 new services: one to call the update manually once, one to play the lost iphone sound on a device and one to change the interval.

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
icloud:
  bart:
    username: <my username>
    password: <my password>
    ignored_devices:
      - iphonevanbvb
      - ipodvanbart
    google_maps_travel_time:
      iphonebart: sensor.bart_naar_huis
    cookie_directory: 'C:\temp'
  sharon:
    username: <gf's username>
    password: <gf's password>
```

Bart274 committed:
change icloud device_tracker into new platform

Bart274 committed:
updated docstrings

Bart274 committed:
fix some travis build errors

Bart274 committed:
fix travis errors

Bart274 committed:
pylint fixes

Bart274 committed:
remove import pyicloud from global scope

Bart274 committed:
remove unused imports + disable pylint warnings

Bart274 committed:
removed manual_update and cleaned up code

Bart274 committed:
fixes

persandstrom commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,952 @@
+                to_dt = from_dt + timedelta(days=7)
+                events = self.api.calendar.events(from_dt, to_dt)
+                new_events = sorted(events, key=self.get_key)
+                starttime = None
```
There is no need set these values to ```None```. That applies to several places in the file.

Bart274 replied:
@persandstrom hmm, I'm used to doing that :-) these are local variables, are they cleared automatically when I call a function again?

persandstrom replied:
The thing I said about loop is not true in python. Working in too many languages at once at too little in each😀

Bart274 committed:
fix build error

Bart274 committed:
removed some None assigments

Bart274 committed:
some more unnecessary assigments removed

Bart274 committed:
removed icloud.py from coverage

persandstrom commented:
If Event is a standard dict it will raise an exception if the key does not
exist.

On Oct 19, 2016 4:13 PM, "Bart274" <notifications@github.com> wrote:

*>*

*> @Bart274* commented on this pull request.
>
> ________________________________
>
> In homeassistant/components/icloud.py
<https://github.com/home-assistant/home-assistant/pull/3946>:
>
> > + cookie_directory=self.cookiedir, + verify=True) + except
PyiCloudFailedLoginException as error: + _LOGGER.error('Error logging into
iCloud Service: %s', + error) + + if self.api is not None: +
self.api.authenticate() + for devicename in self.devices: +
self.devices[devicename].keep_alive() + if self.getevents: + from_dt =
dt_util.now() + to_dt = from_dt + timedelta(days=7) + events =
self.api.calendar.events(from_dt, to_dt) + new_events = sorted(events,
key=self.get_key) + starttime = None
> <https://github.com/persandstrom>
> @persandstrom <https://github.com/persandstrom> will this starttime =
event['startDate'] put starttime to None if the event doesn't have a
startDate?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
<https://github.com/home-assistant/home-assistant/pull/3946>, or mute the
thread
<https://github.com/notifications/unsubscribe-auth/ABzM_dLDD5ctapXhLgZWk72NQDIFG65bks5q1iV_gaJpZM4KargI>
.

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+ENTITY_ID_FORMAT_DEVICE = DOMAIN2 + '.{}'
+ENTITY_ID_FORMAT_EVENT = DOMAIN3 + '.{}'
+
+DEVICESTATUSSET = ['features', 'maxMsgChar', 'darkWake', 'fmlyShare',
```
Can you maybe get this from `pyicloud`?

Bart274 replied:
unfortunately, they're not in pyicloud, someone from that repo told me all the possible values, but it's not in the repo itself

kellerza replied:
Fine to keep it here for now, but might be good to also submit a PR upstream

Bart274 replied:
How would you suggest I make the PR upstream? I could only add each name to the pyicloud repo, because the set I defined here has all the available attributes for the iDevices, but other programs might not want all this data, so this can't be a fixed set in there :s

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+        if CONF_USERNAME not in account_config:
+            _LOGGER.error("Missing username for account %s", account)
+            continue
+        if CONF_PASSWORD not in account_config:
```
Please move all these checks to voluptuous CONFIG_SCHEMA

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+
+def setup(hass, config):  # pylint: disable=too-many-locals,too-many-branches
+    """Set up the iCloud Scanner."""
+    if config.get(DOMAIN) is None:
```
This should not happen (especially with the correct voluptuous schema)

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+        # Get the username and password from the configuration
+        username = account_config.get(CONF_USERNAME)
+        password = account_config.get(CONF_PASSWORD)
+        cookiedirectory = account_config.get(CONF_COOKIEDIRECTORY, None)
```
The default should be `None`

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+
+        ignored_devices = []
+        if 'ignored_devices' in account_config:
+            ignored_dev = account_config.get('ignored_devices')
```
`.get('ignored_devices', [])` and then you dont need the "if"
Since this is from config, it might be good to create a constant that you share between the voluptuous schema and this get

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+
+        googletraveltime = {}
+        if 'googletraveltime' in account_config:
+            gttconfig = account_config.get('googletraveltime')
```
Same here please

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+                ATTR_DEVICESTATUS: self._devicestatus,
+                ATTR_LOWPOWERMODE: self._lowpowermode,
+                ATTR_BATTERYSTATUS: self._batterystatus,
+                ATTR_GOOGLE_MAPS_TRAVEL_TIME: self._googletraveltime
```
This probably does not belong here

Bart274 replied:
what do you mean?

kellerza replied:
On line 246 in this branch of `if` the `self._googletraveltime` var is `None`

Bart274 replied:
ooh right :-D

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+        self._request_interval_seconds = 60
+        self._interval = 1
+        self.api = icloudobject.api
+        self._distance = None
```
Maybe it will be better to build a dict, say `self._attrs` with all these properties and then you can simply return it in state_attributes?

Bart274 replied:
I have no idea where I would create the dict. Do I need to recreate it with each state update or can I just create the dict ones and will the values be updated automatically?

kellerza replied:
You build it later on where you receive the values...

Here you only create the dict instead of all these `=None` assignments
```python
self._attrs = {}
```

You populate it as you interrogate iCloud on other places...
```
self._attrs[ATTR_DEVICENAME] = the_value
# and for almost all of your ATTR_s
```

and then the complete `state_attributes` function becomes:
```python
def state_attributes(self):
    """Return the friendlyname of the iDevice."""
    return self._attrs
```

Bart274 replied:
I see, I'll make the change :-)

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+        currentminutes = dt_util.now().hour * 60 + dt_util.now().minute
+        if currentminutes % self._interval == 0:
+            self.update_icloud()
+        elif self._interval > 10 and currentminutes % self._interval == 2:
```
Instead of =2 and =4, rather `in [2, 4]`?

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,946 @@
+            if nextev:
+                self._remaining = self._starttime - tempnow
+                remainingdays = self._remaining.days
+                remainingseconds = (self._starttime.hour * 3600 +
```
Is it not possible to do math and comparisons directly on a datetime?

Bart274 replied:
That's the thing, when I compared the datetimes, the remaining time was always 45 minutes off. :s

kellerza replied:
Strange, but this is the type of thing unit tests should clarify...

balloob replied:
I agree with @kellerza . Code like this is unmaintainable and instant tech debt.

Bart274 replied:
I'll try again with the default datetime comparisons, maybe it works again now

kellerza reviewed (changes requested):
Thanks @Bart274!

Do you want to add custom UI elements for the idevice Entity and displaying calendar events?

Bart274 commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -1,100 +0,0 @@
-    })
-
-
-def setup_scanner(hass, config, see):
```
Ok,  I'll revert the delete :-)

Bart274 commented on .coveragerc:
```diff
@@ -27,6 +27,9 @@ omit =
     homeassistant/components/envisalink.py
     homeassistant/components/*/envisalink.py
+    
+    homeassistant/components/icloud.py
```
I wouldn't be able to test everything, only the logic from the events and the idevices, but I wouldn't be able to test the icloud api itself. Is that ok?

Bart274 committed:
add voloptuous

Bart274 committed:
pylint fix

Bart274 committed:
Re-add old iCloud Tracker + Use slugify for device names

Bart274 committed:
removed unused import `re`

Bart274 commented:
closed and reopened to trigger travis :-)

Bart274 commented:
hmm damnit, all my changes made my component stop running. it gets loaded without an issue, but when HA initializes the states of the device_trackers, it just hangs without any errors

Bart274 committed:
Fixed config validation

Bart274 committed:
update services.yaml for icloud services

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+                                                            duration,
+                                                            title,
+                                                            tzone,
+                                                            location)
```
A bit more (compact) & readable formatting here would be:

```python
self.currentevents[guid].keep_alive(
    starttime, endtime, duration, title, tzone, location)
```
Open the bracket, immediately newline and 4 spaces indented...

kellerza replied:
(also applicable in other places)

Bart274 replied:
ok, will change it :-)

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+            from pyicloud.exceptions import PyiCloudNoDevicesException
+
+            try:
+                # Loop through every device registered with the iCloud account
```
Not sure this comment makes sense here

kellerza commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+            ATTR_DURATION: self._duration,
+            ATTR_REMAINING: self._remainingtext,
+            ATTR_LOCATION: self._location,
+            ATTR_FRIENDLY_NAME: self._type
```
Here we can use the same `dict` approach I suggested above.

The main reason for this is that all these attributes are only displayed through the `state_attributes` function. We might as well put it in the correct format while we read the data

kellerza commented:
It would be great to see a bit more use of Python's `datetime.*` & `timedelta` standard library functions instead of reinventing The Wheel of Time...

balloob commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+DOMAIN3 = 'ievent'
+
+ENTITY_ID_FORMAT_ICLOUD = DOMAIN + '.{}'
+ENTITY_ID_FORMAT_DEVICE = DOMAIN2 + '.{}'
```
You are only allowed to publish entities under your own domain

balloob commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+ATTR_LOWPOWERMODE = 'low_power_mode'
+ATTR_BATTERYSTATUS = 'battery_status'
+ATTR_LOCATION = 'location'
+ATTR_FRIENDLY_NAME = 'friendly_name'
```
You should never have to use this. Just have your entities have a name and `update_ha_state` will take care of this.

balloob commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+            }
+        else:
+            return {
+                ATTR_DEVICENAME: self.devicename,
```
There is no reason to have all this duplication. Please consolidate.

balloob commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+    return True
+
+
+class IDevice(Entity):  # pylint: disable=too-many-instance-attributes
```
There shouldn't be ever 1 component with 3 different entities. Spread it out. Keep the device tracker platform a device tracker platform but have it depend on this code, hook it in .

Bart274 commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+        tempdays = floor(self._remaining / 1440)
+        temphours = floor((self._remaining % 1440) / 60)
+        tempminutes = floor(self._remaining % 60)
+        self._remainingtext = (str(tempdays) + "d " +
```
Why does that make no sense? How else would I update the remaining time of this entity?
The removal of the entity, I might change with the track_point_in_time function right?

Bart274 commented on homeassistant/components/icloud.py:
```diff
@@ -0,0 +1,956 @@
+            from pyicloud.exceptions import PyiCloudFailedLoginException
+            try:
+                # Attempt the login to iCloud
+                self.api = PyiCloudService(self.username,
```
I'm not sure how to do this?

Bart274 commented:
@kellerza this is the reason why I changed the code to calculate the remaining time.
When I use the timedelta itself for a current event, it gives me the correct remaining time (I just need to check again with the timezones, that is still a difference, but I think I know how to solve that).
But when I check the timedelta for a future event, the remaining time is just wrong. Look at the attached image. The screenshot is taking at 14:32, the start time is 20:00 (in a few days, but the remaining days is correct), but the amount of minutes remaining is off. Even with the timezone difference, there still is a mistake of 18 minutes.

![timedifference](https://cloud.githubusercontent.com/assets/1118630/19598641/24161bc0-979c-11e6-8ed6-26565e0337db.JPG)

Bart274 committed:
clean up code + remove calendar part for now

Bart274 commented:
Ok, for now, I have removed the calendar part of this component for this PR. I'm going to focus first on replacing the old device_tracker with the new device_tracker part and I can always add the calendar part back in the future.

Bart274 committed:
removed unused code

Bart274 committed:
fix indentation and remove unused imports

Bart274 commented:
@kellerza thanks for the advice! I'll look at it on monday and see if I can use that 😊

kellerza commented:
No problem, this will fit right in with the HASS model!

For other components (not device_tracker) you have `discovery.load_platform` to perform this type of function, but lets hope you can make it work with `see`

Bart274 committed:
Updated component to only have one entity type

This commit needs
https://github.com/home-assistant/home-assistant/pull/4023 to be merged
first. There is an error in the See service that has been fixed for this
PR.

Bart274 committed:
pylint errors fixed

Bart274 committed:
forgot to remove some code

Bart274 committed:
if the interval changes, update the device_tracker entity

Bart274 commented:
@balloob & @kellerza what do you think of this latest version? :-)
Just one issue, someone is using the version and he noticed his device_tracker entities aren't showing up.
I'm going to test this as soon as I have access to my server again. Somehow I think the entities aren't added to the known_devices.yaml file because my component isn't a device_tracker?

Bart274 committed:
fixed error in setinterval service with no devicename given

Bart274 committed:
support for 2FA

Bart274 committed:
fix travis errors

Bart274 committed:
removed spaces from coverage

Bart274 commented:
I'm closing this PR in favor of a new one I'm going to make in a bit. It will no longer add icloud as a new platform, but instead it will upgrade the current device_tracker version.
